### DISCUSSION TITLE
Consistent OCA canonical form

### DISCUSSION THREAD

**Discussion by blelump** [the-human-colossus-foundation/oca-spec#69] [General]
[Deterministic mechanics](https://hackmd.io/@bYQK_qO_RLa70okz8n7TQg/ry30vwKx6) is a must-have consequence of immutable data structures. However, to ensure a piece of data is available in a readable form, we must guarantee the proper ordering of the readable chunks so that when it comes to applying deterministic mechanics, the ordering is always the same. The ordering of whatever we feed the one-way hash function to calculate the digest must conform to the following rules:
1. The so-called Mechanics Bundle is an object that has the following attributes in this order: `v`, `d`, `capture_base`, `overlays`.
1. `capture_base` has the following attributes in this order: `v`, `type`, `attributes`. 
    - The `attributes` array has ASC order. 
    - _The currently existing `classification` and `flagged_attributes` shall be extracted into overlays._
1. `overlays` is an object whose ordering must adhere to https://www.rfc-editor.org/rfc/rfc8785#section-3.2.3 spec (with minor exceptions, see below). Each value, the actual overlay, however, must have the following attributes in this order: `d`, `type` that constitutes the overlay base. The remaining, overlay-specific attributes are then amended to the overlay base with ASC ordering.
    - any nested non-primitive overlay attribute must adhere to https://www.rfc-editor.org/rfc/rfc8785#section-3.2.3
    - overlays, i.e., label, entry, information, that cardinality in the bundle is `> 1`, are sorted in ASC order by the `d` attribute

\* ASC order is lexicographic (alphabetical) order.

**Comment by carlyh-micb** [comment]
The TC is coming up with ideas that are not shared with the community. What is a mechanics bundle? Are there any examples, code, documentation? It doesn't need to be fully formed before it is presented, it would be a stronger idea with contributions from the community as well as it is developed.

**Reply by blelump** [reply]
Briefly introduced [here](https://github.com/the-human-colossus-foundation/oca-spec/issues/53#issuecomment-2346527857) but demands a separate discussion. The mechanics bundle is the current bundle.

**Reply by carlyh-micb** [reply]
"The mechanics bundle is the current bundle." - very important detail that hasn't been communicated. 
Can we switch back to Semantic Versioning for the version numbers?
Does the TC have a draft of a the proposed structure, assuming the mechanics bundle is not the only one?
ADC is moving forward with OCA Package to incorporate our necessary overlays and will present updates on this work at the DSWG. The prototype is working now. We use semantic versioning, SAID calculations harmonize overlays to the syntax of OCAv1.0.

**Reply by blelump** [reply]
> Can we switch back to Semantic Versioning for the version numbers?

WDYM?

> Does the TC have a draft of a the proposed structure, assuming the mechanics bundle is not the only one?

We'll publish it as soon as it's done.

**Reply by carlyh-micb** [reply]
> We'll publish it as soon as it's done.

Let's work together more as a community to build something that consistently aligns with user needs and gets input into the designs? Publish a draft/work in progress and have comments.

**Reply by pknowl** [reply]
The term _"Mechanics Bundle"_ implies a focus on actions, events, and the dynamics of how things operate, which seems misaligned with the structural and morphological nature of what Michal is describing. Since this seems to be more about organising and structuring data objects, I'm sure we can come up with a more granularly accurate term to describe this concept, i.e. terms that better reflect the idea of handling the composition and order of objects and overlays within a digital format, rather than focusing on the mechanics or processes acting upon them.

**Reply by pknowl** [reply]
_"Stratification Bundle"_ might work well. It suggests the idea of layers or strata, emphasising the ordered and structured nature of the bundle's components (capture base, overlays, etc.). This term captures the essence of organisation and layering without the process-oriented connotations of "mechanics." It aligns with the idea that each layer is distinct but forms part of a cohesive whole, which seems fitting for what Michal describes.

**Reply by pknowl** [reply]
"deterministic stratification" ...
https://tuprints.ulb.tu-darmstadt.de/2308/1/dissertation_girschick.pdf

**Reply by pknowl** [reply]
Delving deeper ...

Question to ChatGPT:

When constructing a house, various components like wood beams, steel girders of specific dimensions and strength, and windows are required. What would be an appropriate collective term to refer to these essential construction elements?

Answer:

A good enveloping name for the components required to build a house, like wood beams, steel girders, and windows, could be "structural elements". The term is broad enough to cover the variety of components needed, ranging from foundational materials like steel and wood to finishing elements like windows and doors.

"Structural elements" emphasizes their role in providing the framework and integrity of the house.

@blelump / @mitfik - Semantically, I prefer _"Structural Bundle"_ over _"Mechanics Bundle"_.

- **Structural elements**: These define the underlying framework or schema that supports and organises data. In digital contexts, this could include data models, formats, or protocols that provide the structure for information storage and retrieval (e.g., JSON schema, database structures, or XML elements).

- **Presentation elements**: These are concerned with how data or content is displayed or rendered to users. They define the visual or interactive layout and include things like CSS, HTML, or UI frameworks that shape the user experience but do not alter the core data.

**Reply by blelump** [reply]
Structural Bundle sounds good.

**Comment by pknowl** [comment]
Other points of note ... 

1. What was the rationale behind sorting overlays by the "d" attribute instead of the overlay name, which is more intuitive for humans to read and the first thing available for sorting?

2. Why does the current approach utilise different canonicalization methods for different parts to calculate SAIDs, when [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785) provides a simpler and more standardised method?

3. Would the TC be open to simplifying canonicalization by applying RFC 8785 and eliminating the need for complex methods that are unnecessary for human readability and processing?

**Reply by blelump** [reply]
1. There's no such statement in the above proposal. What it states for overlays that have cardinality > 1 (i.e., label) is to sort them by  `d` attribute to ensure unambiguous canonical form. In other words, if OCA Bundle provides 15 translations, that is, 15 label overlays, they must be unambiguously ordered, and considering future support for community overlays, `d` is the common denominator addressing ordering for any overlay with cardinality > 1. Note this is ordering a vector of overlays of the same type.
2. Custom ordering used in the OCA Bundle adds essential benefits that are critical to the whole ecosystem:
    1. having `v` (version) as the first attribute enables reading the version without deserializing the Bundle (note it also works for other representations). In particular, the version can be read directly from the serialized vector of bytes, i.e., JSON string. This is a highly effective technique for further decision-making processes on handling different versions, if at all, without premature deserialization. A production-ready system must embrace changes to keep backward and forward compatibility. For data sent over the wire with a specific structure, embrace structure evolution over time. This is achievable by versioning, and effectively getting the version is a vital topic to address. We also observe the usefulness of such an approach in production systems where Bundles reach several megabytes. Deserializing such structures at scale is a time and resource cost.
    2. Having `d` as the second attribute has similar reasoning as `v` plus an additional advantage useful for storage. `d` is the unique identifier of the Bundle and its parts, making it not only a mechanism for deterministic verification but also an identifier in any type of storage, including caches, key-value storages, etc. It's therefore crucial to have `d` unambiguously given without deserializing the actual data structure. 
3. thinking of a data structure is not only how it serializes to be sent over the wire. More importantly, how it becomes consumable and accessible in various use cases. With [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785) in place, both `v` and `d` become inaccessible as they are at unknown positions in the serialized representation. While searching for specific characters is doable, it's less effective and error-prone, especially in nested structures (i.e., OCA Bundle). [RFC 8785](https://www.rfc-editor.org/rfc/rfc8785) neither discusses nor addresses these aspects – it's beyond its responsibility. Therefore, we embrace custom solutions that serve more value to the ecosystem in the long term. It's also our responsibility to consider more essential topics beyond serialization, where serialization imposes (narrows) specific solutions. With what we introduced a long time ago in the OCA Bundle, it's possible to:
    1. enable effective decision-making business processes 
    2. use cache mechanisms for the Bundles, i.e., where no OCA Repo is ever involved
    3. verify the data integrity of the data structure by directly manipulating its serialized form and executing a one-way hash function on it to verify the integrity.

The optimizations we introduced by such custom ordering benefited various corners of the ecosystem. Last but not least, the serialization order does not anyhow impact the actual data structure. It's solely the ordering in a given representation.

What seems to be missing in these questions is that a custom representation may bring more human value to such mechanical creatures by ensuring the attribute ordering is constant and always the same. This is in contrast to RFC 8785, where it is always lexicographic. It happens to study a serialized Bundle, and knowing by heart where to look for specific information makes life easier at the end of the day.
